### ISSUE TITLE
_prepare_4d_causal_attention_mask mask inversion should work boolean masks

### ISSUE THREAD

**Issue by JackCai1206** [huggingface/transformers#32113]
### Feature request

https://github.com/huggingface/transformers/blob/0fdea8607d7e01eb0e38a1ebeb7feee30a22f0cf/src/transformers/modeling_attn_mask_utils.py#L332

Here we can just assume the user provides a tensor of 0s and 1s right? In that case, the inversion can be done like `inverted_mask = ~(attention_mask.to(torch.bool))`

### Motivation

There are two advantages I found with doing the conversion assuming the input is bool
* Its slightly faster 
* When using a cached huggingface dataset to generate the mask, it is way more space-efficient to store the masks as bool type. 

### Your contribution

`inverted_mask = ~(attention_mask.to(torch.bool))`

**Comment by amyeroberts** [comment]
cc @ArthurZucker

**Comment by ArthurZucker** [comment]
I would need to check where this is used, but it's not used everywhere (especially llama like) cc @zucchini-nlp we saw a recent comment about that! 
WOuld you like to open a PR let's see if this works?

**Comment by zucchini-nlp** [comment]
Yeah, I think soon we will not be using that anymore and rely on `update_causal_mask` method similar to llama. The comment we saw was about accepting attn as boolean type for `forward()` call, and that will definitely break things in VLMs (maybe not LLMs). I guess this issue is quite different but we can see how it will fit :)

**Comment by smpanaro** [comment]
I want to +1 this since it caught me off guard. I was passing a boolean mask which worked on CPU since it used `scaled_dot_product_attention` which supports both boolean and inverted masks. Switching to GPU caused me to switch attention implementations and lose that behavior silently.

There used to be an error for this case, but it was removed:
https://github.com/huggingface/transformers/pull/32227/files#diff-0ebada593d7651a4bd67df9568471123d78edee35a2f0b8bb44751a192e5a4deL777-L781

Either adding it back, or adding support for boolean masks would be great.
